pkg/handler/common/kubeconfig.go

```diff
@@ -286,6 +294,190 @@ func CreateOIDCKubeconfigEndpoint(ctx context.Context, projectProvider provider.
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+		if err := createKubeconfigSecret(ctx, client, oidcKubeCfg, claims.Email); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+		return rsp, nil
+	}
+
+	// PHASE initial handles request from the end-user that wants to authenticate
+	// and kicksoff the process of kubeconfig generation
+	if req.phase != initialPhase {
+		return nil, utilerrors.NewBadRequest(fmt.Sprintf("bad request unexpected phase %d, expected phase %d, did you forget to set the phase while decoding the request?", req.phase, initialPhase))
+	}
+
+	rsp := createOIDCKubeconfigRsp{}
+	scopes := []string{"openid", "email"}
+	if oidcCfg.OfflineAccessAsScope {
+		scopes = append(scopes, "offline_access")
+	}
+
+	// Override default redirect uri
+	oidcIssuerVerifier.SetRedirectURI(getRedirectURL(req.request))
+
+	// pass nonce
+	nonce := rand.String(rand.IntnRange(10, 15))
+	rsp.nonce = nonce
+	rsp.secureCookieMode = oidcCfg.CookieSecureMode
+
+	oidcState := OIDCState{
+		Nonce:     nonce,
+		ClusterID: req.ClusterID,
+		ProjectID: req.ProjectID,
+		UserID:    req.UserID,
+	}
+	rawState, err := json.Marshal(oidcState)
+	if err != nil {
+		return nil, err
+	}
+	encodedState := base64.StdEncoding.EncodeToString(rawState)
+	urlSafeState := url.QueryEscape(encodedState)
+	rsp.authCodeURL = oidcIssuer.AuthCodeURL(urlSafeState, oidcCfg.OfflineAccessAsScope, scopes...)
+
+	return rsp, nil
+}
+
+func createKubeconfigSecret(ctx context.Context, client ctrlruntimeclient.Client, config *clientcmdapi.Config, email string) error {
+	// encode email address to unique ID for the secret name
+	hasher := md5.New()
+	hasher.Write([]byte(email))
+	kubeconfigSecretName := hex.EncodeToString(hasher.Sum(nil))
+	kubeconfig, err := clientcmd.Write(*config)
+	if err != nil {
+		return err
+	}
+
+	namespacedName := types.NamespacedName{Namespace: resources.KubeSystemNamespaceName, Name: kubeconfigSecretName}
+
+	existingSecret := &corev1.Secret{}
+	if err := client.Get(ctx, namespacedName, existingSecret); err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to probe for secret %v: %w", namespacedName, err)
+	}
+
+	secretData := map[string][]byte{
+		resources.KubeconfigSecretKey: kubeconfig,
+	}
+
+	// return if already exists
+	if existingSecret.Name != "" {
+		return nil
+	}
+
+	return createSecret(ctx, client, kubeconfigSecretName, email, secretData)
+}
+
+func createSecret(ctx context.Context, client ctrlruntimeclient.Client, name, email string, secretData map[string][]byte) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   resources.KubeSystemNamespaceName,
+			Annotations: map[string]string{"user": email},
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: secretData,
+	}
+	return client.Create(ctx, secret)
+}
+
+func getRedirectURL(r *http.Request) string {
+	scheme := "http"
+	if r.TLS != nil {
+		scheme = "https"
+	}
```

Comment:
This doesn't seem to work, at least it doesn't here https://github.com/kubermatic/kubermatic/blob/master/pkg/handler/v2/kubernetes-dashboard/login.go#L152

And the failure seems right because there is no mTLS and the API inside the pod is served over HTTP; 0.0.0.0:8080
.

Reply:
good point. I've changed it. We can rely on the current redirect URI and just change the path of the endpoints